### PR TITLE
[PollyTTS] voice service

### PR DESCRIPTION
**Pull request by hillmanr**
PollyTTS Text-to-Speech

PollyTTS is a TTS voice service utilizing a Internet based TTS service provided by amazon called Polly.  There are five servers set in various geographic locations.   It requires a set of API keys provided by Amazon to get access to the service.  Provides multiple languages and multiple voices for each language. The free tier includes 5 million characters per month for the first 12 months. Then it's $4 for a million characters after the one year.

For more information, see https://aws.amazon.com/polly/

[prior discussion](https://community.openhab.org/t/amazon-polly-tts-voice-service/29699)

### CODE REVIEW

**Comment by kubawolanin** [comment]
Could you add the addon to the parent `pom.xml`, so it's captured by the distro?
https://github.com/openhab/openhab2-addons/blob/master/addons/voice/pom.xml

**Comment by kubawolanin** [comment]
BTW. For anyone interested - latest build of this binding can be found here:
https://openhab.jfrog.io/openhab/libs-pullrequest-local/org/openhab/voice/org.openhab.voice.pollytts/2.1.0-SNAPSHOT/

**Comment by martinvw** [comment]
@hillmanr Please note that, because it's marked [WIP], I do and did not review this PR completely yet.

If it's finished and ready to review please update the title and mention me so I can remove the "Work in progress" label.

**Comment by kubawolanin** [comment]
@hillmanr any news on your side? I'm really looking forward to see it in the distro :-)

BTW. It's worth noting that this PR closes #1439

**Comment by kubawolanin** [comment]
@hillmanr I've encountered an issue with your Add-on.
I'm on openHAB build # 1099 and on openHAB runtime startup I get the following errors:

```
2017-11-26 13:57:36.086 [ERROR] [ce.pollytts.internal.PollyTTSService] - Failed to activate PollyTTS: For input string: "0.0"
java.lang.NumberFormatException: For input string: "0.0"
	at java.lang.NumberFormatException.forInputString(NumberFormatException.java:65) ~[?:?]
	at java.lang.Integer.parseInt(Integer.java:580) [?:?]
	at java.lang.Integer.parseInt(Integer.java:615) [?:?]
	at org.openhab.voice.pollytts.internal.PollyTTSService.modified(PollyTTSService.java:88) [10:org.openhab.voice.pollytts:2.1.0.201706261335]
	at org.openhab.voice.pollytts.internal.PollyTTSService.activate(PollyTTSService.java:70) [10:org.openhab.voice.pollytts:2.1.0.201706261335]
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[?:?]
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62) ~[?:?]
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[?:?]
	at java.lang.reflect.Method.invoke(Method.java:498) ~[?:?]
	at org.apache.felix.scr.impl.inject.BaseMethod.invokeMethod(BaseMethod.java:229) [50:org.apache.felix.scr:2.0.12]
	at org.apache.felix.scr.impl.inject.BaseMethod.access$500(BaseMethod.java:39) [50:org.apache.felix.scr:2.0.12]
	at org.apache.felix.scr.impl.inject.BaseMethod$Resolved.invoke(BaseMethod.java:650) [50:org.apache.felix.scr:2.0.12]
	at org.apache.felix.scr.impl.inject.BaseMethod.invoke(BaseMethod.java:506) [50:org.apache.felix.scr:2.0.12]
	at org.apache.felix.scr.impl.inject.ActivateMethod.invoke(ActivateMethod.java:307) [50:org.apache.felix.scr:2.0.12]
	at org.apache.felix.scr.impl.inject.ActivateMethod.invoke(ActivateMethod.java:299) [50:org.apache.felix.scr:2.0.12]
	at org.apache.felix.scr.impl.manager.SingleComponentManager.createImplementationObject(SingleComponentManager.java:298) [50:org.apache.felix.scr:2.0.12]
	at org.apache.felix.scr.impl.manager.SingleComponentManager.createComponent(SingleComponentManager.java:109) [50:org.apache.felix.scr:2.0.12]
	at org.apache.felix.scr.impl.manager.SingleComponentManager.getService(SingleComponentManager.java:906) [50:org.apache.felix.scr:2.0.12]
	at org.apache.felix.scr.impl.manager.SingleComponentManager.getServiceInternal(SingleComponentManager.java:879) [50:org.apache.felix.scr:2.0.12]
	at org.apache.felix.scr.impl.manager.AbstractComponentManager.activateInternal(AbstractComponentManager.java:749) [50:org.apache.felix.scr:2.0.12]
	at org.apache.felix.scr.impl.manager.AbstractComponentManager.enableInternal(AbstractComponentManager.java:675) [50:org.apache.felix.scr:2.0.12]
	at org.apache.felix.scr.impl.manager.AbstractComponentManager.enable(AbstractComponentManager.java:430) [50:org.apache.felix.scr:2.0.12]
	at org.apache.felix.scr.impl.manager.ConfigurableComponentHolder.enableComponents(ConfigurableComponentHolder.java:657) [50:org.apache.felix.scr:2.0.12]
	at org.apache.felix.scr.impl.BundleComponentActivator.initialEnable(BundleComponentActivator.java:341) [50:org.apache.felix.scr:2.0.12]
	at org.apache.felix.scr.impl.Activator.loadComponents(Activator.java:390) [50:org.apache.felix.scr:2.0.12]
	at org.apache.felix.scr.impl.Activator.access$200(Activator.java:54) [50:org.apache.felix.scr:2.0.12]
	at org.apache.felix.scr.impl.Activator$ScrExtension.start(Activator.java:265) [50:org.apache.felix.scr:2.0.12]
	at org.apache.felix.utils.extender.AbstractExtender.createExtension(AbstractExtender.java:254) [50:org.apache.felix.scr:2.0.12]
	at org.apache.felix.utils.extender.AbstractExtender.modifiedBundle(AbstractExtender.java:227) [50:org.apache.felix.scr:2.0.12]
	at org.apache.felix.utils.extender.AbstractExtender.addingBundle(AbstractExtender.java:187) [50:org.apache.felix.scr:2.0.12]
	at org.osgi.util.tracker.BundleTracker$Tracked.customizerAdding(BundleTracker.java:469) [?:?]
	at org.osgi.util.tracker.BundleTracker$Tracked.customizerAdding(BundleTracker.java:415) [?:?]
	at org.osgi.util.tracker.AbstractTracked.trackAdding(AbstractTracked.java:256) [?:?]
	at org.osgi.util.tracker.AbstractTracked.track(AbstractTracked.java:229) [?:?]
	at org.osgi.util.tracker.BundleTracker$Tracked.bundleChanged(BundleTracker.java:444) [?:?]
	at org.eclipse.osgi.internal.framework.BundleContextImpl.dispatchEvent(BundleContextImpl.java:903) [?:?]
	at org.eclipse.osgi.framework.eventmgr.EventManager.dispatchEvent(EventManager.java:230) [?:?]
	at org.eclipse.osgi.framework.eventmgr.ListenerQueue.dispatchEventSynchronous(ListenerQueue.java:148) [?:?]
	at org.eclipse.osgi.internal.framework.EquinoxEventPublisher.publishBundleEventPrivileged(EquinoxEventPublisher.java:213) [?:?]
	at org.eclipse.osgi.internal.framework.EquinoxEventPublisher.publishBundleEvent(EquinoxEventPublisher.java:120) [?:?]
	at org.eclipse.osgi.internal.framework.EquinoxEventPublisher.publishBundleEvent(EquinoxEventPublisher.java:112) [?:?]
	at org.eclipse.osgi.internal.framework.EquinoxContainerAdaptor.publishModuleEvent(EquinoxContainerAdaptor.java:156) [?:?]
	at org.eclipse.osgi.container.Module.publishEvent(Module.java:476) [?:?]
	at org.eclipse.osgi.container.Module.start(Module.java:467) [?:?]
	at org.eclipse.osgi.container.ModuleContainer$ContainerStartLevel.incStartLevel(ModuleContainer.java:1620) [?:?]
	at org.eclipse.osgi.container.ModuleContainer$ContainerStartLevel.incStartLevel(ModuleContainer.java:1599) [?:?]
	at org.eclipse.osgi.container.ModuleContainer$ContainerStartLevel.doContainerStartLevel(ModuleContainer.java:1571) [?:?]
	at org.eclipse.osgi.container.ModuleContainer$ContainerStartLevel.dispatchEvent(ModuleContainer.java:1514) [?:?]
	at org.eclipse.osgi.container.ModuleContainer$ContainerStartLevel.dispatchEvent(ModuleContainer.java:1) [?:?]
	at org.eclipse.osgi.framework.eventmgr.EventManager.dispatchEvent(EventManager.java:230) [?:?]
	at org.eclipse.osgi.framework.eventmgr.EventManager$EventThread.run(EventManager.java:340) [?:?]
```

I'd appreciate any help.
Thanks

**Comment by kubawolanin** [comment]
@hillmanr the add-on doesn't work with latest SNAPSHOT distro at all. See https://github.com/eclipse/smarthome/issues/4632

Thank you.

File: addons/voice/org.openhab.voice.pollytts/ESH-INF/config/config.xml [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+                  <default>eu-west-1</default>
+               </parameter>
+	       <parameter name="audioFormat" type="text" required="false">
+                  <label>Optional User specified audio format</label>
+                  <description>The user can override the system default audio format with their prefered option. 
+                   "MP3"" and "OGG" are the only audio formats that are supported. </description>
+                  <options>
+                      <option value="OGG">OGG audio format</option>
+                      <option value="MP3">MP3 format</option>
+                      <option value="disable">Use system selected audio format</option>
```
[the remark this answers is not in the document]
Comment (hillmanr) [reply addons/voice/org.openhab.voice.pollytts/ESH-INF/config/config.xml]: agree. done

File: addons/voice/org.openhab.voice.pollytts/src/main/java/org/openhab/voice/pollytts/internal/PollyTTSService.java [hunk 2]
```diff
@@ -0,0 +1,251 @@
...
+        modified(config);
+    }
+
+    protected void modified(Map<String, Object> config) {
+        try {
+            boolean successful = false;
+            if (config != null) {
+                PollyClientConfig polly = new PollyClientConfig();
+                String val = null;
+                val = config.containsKey(CONFIG_ACCESS_KEY) ? config.get(CONFIG_ACCESS_KEY).toString() : null;
```
[the remark this answers is not in the document]
Comment (hillmanr) [reply addons/voice/org.openhab.voice.pollytts/src/main/java/org/openhab/voice/pollytts/internal/PollyTTSService.java]: removed

File: addons/voice/org.openhab.voice.pollytts/src/main/java/org/openhab/voice/pollytts/internal/PollyTTSService.java [hunk 3]
```diff
@@ -0,0 +1,251 @@
...
+                polly.setAudioFormat(val);
+                val = config.containsKey(CONFIG_EXPIRE) ? config.get(CONFIG_EXPIRE).toString() : "0";
+                polly.setExpireDate(Integer.parseInt(val));
+                successful = polly.initPollyServiceInterface();
+            }
+            if (successful) {
+                pollyTssImpl = initVoiceImplementation();
+                audioFormats = initAudioFormats();
+                voices = initVoices();
+                logger.info("PollyTTS cfg data loaded and static elements (re)initialized");
```
[the remark this answers is not in the document]
Comment (hillmanr) [reply addons/voice/org.openhab.voice.pollytts/src/main/java/org/openhab/voice/pollytts/internal/PollyTTSService.java]: cleaned up all log messages. to correct log levels

File: addons/voice/org.openhab.voice.pollytts/src/main/java/org/openhab/voice/pollytts/internal/PollyTTSService.java [hunk 4]
```diff
@@ -0,0 +1,251 @@
...
+                logger.info("Using PollyTTS cache folder {}", getCacheFolderName());
+            } else {
+                logger.info("PollyTTS not initialized");
+            }
+
+        } catch (Throwable t) {
```
[the remark this answers is not in the document]
Comment (hillmanr) [reply addons/voice/org.openhab.voice.pollytts/src/main/java/org/openhab/voice/pollytts/internal/PollyTTSService.java]: cleaned exceptions and removed code style warnings

File: addons/voice/org.openhab.voice.pollytts/src/main/java/org/openhab/voice/pollytts/internal/PollyTTSService.java [hunk 5]
```diff
@@ -0,0 +1,251 @@
...
+            logger.error("Failed to activate PollyTTS: {}", t.getMessage(), t);
+        }
+    }
+
+    @Override
+    public Set<Voice> getAvailableVoices() {
+        if (this.voices == null) {
```
[the remark this answers is not in the document]
Comment (hillmanr) [reply addons/voice/org.openhab.voice.pollytts/src/main/java/org/openhab/voice/pollytts/internal/PollyTTSService.java]: I don't understand this convention

File: addons/voice/org.openhab.voice.pollytts/src/main/java/org/openhab/voice/pollytts/internal/PollyTTSService.java [hunk 6]
```diff
@@ -0,0 +1,251 @@
...
+            logger.error("PollyTTS interface never initalized, check congiguration elements");
```
[the remark this answers is not in the document]
Comment (hillmanr) [reply addons/voice/org.openhab.voice.pollytts/src/main/java/org/openhab/voice/pollytts/internal/PollyTTSService.java]: modified as discussed above

File: addons/voice/org.openhab.voice.pollytts/src/main/java/org/openhab/voice/pollytts/internal/PollyTTSService.java [hunk 7]
```diff
@@ -0,0 +1,251 @@
...
+        }
+        return this.voices;
+    }
+
+    @Override
+    public Set<AudioFormat> getSupportedFormats() {
+        return this.audioFormats;
```
[the remark this answers is not in the document]
Comment (hillmanr) [reply addons/voice/org.openhab.voice.pollytts/src/main/java/org/openhab/voice/pollytts/internal/PollyTTSService.java]: The audioFormats are no defined in pollytts these are the system set defined in the 
org.eclipse.smarthome.core.voice.TTSService

File: addons/voice/org.openhab.voice.pollytts/src/main/java/org/openhab/voice/pollytts/internal/PollyTTSService.java [hunk 8]
```diff
@@ -0,0 +1,251 @@
...
+        // only a default voice
+        try {
+            File cacheAudioFile = pollyTssImpl.getTextToSpeechAsFile(text, voice.getLabel(),
+                    getApiAudioFormat(requestedFormat));
+            if (cacheAudioFile == null) {
+                throw new TTSException("Could not read from PollyTTS service");
+            }
+            logger.debug("Audio Stream for '{}' in format {}", text, requestedFormat);
+            AudioStream audioStream = new PollyTTSAudioStream(cacheAudioFile, requestedFormat);
+            // AudioStream audioStream = new PollyTTSAudioStream(cacheAudioFile);
```
[the remark this answers is not in the document]
Comment (hillmanr) [reply addons/voice/org.openhab.voice.pollytts/src/main/java/org/openhab/voice/pollytts/internal/PollyTTSService.java]: removed

File: addons/voice/org.openhab.voice.pollytts/src/main/java/org/openhab/voice/pollytts/internal/cloudapi/CachedPollyTTSCloudImplementation.java [hunk 9]
```diff
@@ -0,0 +1,168 @@
...
+ * @author Robert Hillman - converted to Polly API to utilize unique voice names
+ * @author Robert Hillman - added cache management to delete aged entries
+ */
+public class CachedPollyTTSCloudImplementation extends PollyTTSCloudImplementation {
+
+    private final Logger logger = LoggerFactory.getLogger(CachedPollyTTSCloudImplementation.class);
+
+    private final File cacheFolder;
+
+    public CachedPollyTTSCloudImplementation(String cacheFolderName) {
```
[the remark this answers is not in the document]
Comment (hillmanr) [reply addons/voice/org.openhab.voice.pollytts/src/main/java/org/openhab/voice/pollytts/internal/cloudapi/CachedPollyTTSCloudImplementation.java]: Done

File: addons/voice/org.openhab.voice.pollytts/src/main/java/org/openhab/voice/pollytts/internal/cloudapi/CachedPollyTTSCloudImplementation.java [hunk 10]
```diff
@@ -0,0 +1,168 @@
...
+        }
+    }
+
+    /**
+     * Gets a unique filename for a give text, by creating a MD5 hash of it. It
+     * will be preceded by the voice label.
+     *
+     * Sample: "Robert_00a2653ac5f77063bc4ea2fee87318d3"
+     */
+    String getUniqeFilenameForText(String text, String label) {
```
[the remark this answers is not in the document]
Comment (hillmanr) [reply addons/voice/org.openhab.voice.pollytts/src/main/java/org/openhab/voice/pollytts/internal/cloudapi/CachedPollyTTSCloudImplementation.java]: checked it out, I need a unique name the random technique would not be adequate.

File: addons/voice/org.openhab.voice.pollytts/src/main/java/org/openhab/voice/pollytts/internal/cloudapi/CachedPollyTTSCloudImplementation.java [hunk 11]
```diff
@@ -0,0 +1,168 @@
...
+            while (hashtext.length() < 32) {
+                hashtext = "0" + hashtext;
+            }
+            String fileName = label + "_" + hashtext;
+            return fileName;
+        } catch (UnsupportedEncodingException ex) {
+            // should not happen
+            logger.error("Could not create MD5 hash for '{}'", text, ex);
+            return null;
+        } catch (NoSuchAlgorithmException ex) {
```
[the remark this answers is not in the document]
Comment (hillmanr) [reply addons/voice/org.openhab.voice.pollytts/src/main/java/org/openhab/voice/pollytts/internal/cloudapi/CachedPollyTTSCloudImplementation.java]: removed

File: addons/voice/org.openhab.voice.pollytts/src/main/java/org/openhab/voice/pollytts/internal/cloudapi/PollyClientConfig.java [hunk 12]
```diff
@@ -0,0 +1,142 @@
...
+ * static class so the service interface, Credential setup and voice identification
+ * is only performed on initialization
+ *
+ * @author Robert Hillman - initial interface
+ */
+public class PollyClientConfig {
+
+    private final Logger logger = LoggerFactory.getLogger(PollyClientConfig.class);
+
+    private static String accessKey = null;
```
[the remark this answers is not in the document]
Comment (hillmanr) [reply addons/voice/org.openhab.voice.pollytts/src/main/java/org/openhab/voice/pollytts/internal/cloudapi/PollyClientConfig.java]: agreed

[the remark this answers is not in the document]
Comment (hillmanr) [reply addons/voice/org.openhab.voice.pollytts/src/main/java/org/openhab/voice/pollytts/internal/cloudapi/PollyClientConfig.java]: This class implements a client interface to the amazon polly service. It has
 static data so the service interface, Credential setup and voice identification
 is only performed on initialization. Once initialized the data stays constant and Credential verification is not necessary to repeat. Using Static class variables is the only way I know to do this. If you can recommend an alternate method I can look into it.

File: addons/voice/org.openhab.voice.pollytts/src/main/java/org/openhab/voice/pollytts/internal/cloudapi/PollyClientConfig.java [hunk 13]
```diff
@@ -0,0 +1,142 @@
...
+    private static String secretKey = null;
+    private static String regionVal = null;
+    private static int expireDate = 30;
+    private static String audioFormat = null;
+    private static long today = 0;
+
+    public static AmazonPollyClient polly;
+    public static List<Voice> pollyVoices;
+    // translation function from unique voice label to voice id
+    public static HashMap<String, String> labelToID = new HashMap<String, String>();
```
[the remark this answers is not in the document]
Comment (hillmanr) [reply addons/voice/org.openhab.voice.pollytts/src/main/java/org/openhab/voice/pollytts/internal/cloudapi/PollyClientConfig.java]: deleted

File: addons/voice/org.openhab.voice.pollytts/src/main/java/org/openhab/voice/pollytts/internal/cloudapi/PollyClientConfig.java [hunk 14]
```diff
@@ -0,0 +1,142 @@
...
+        PollyClientConfig.today = today;
+        ;
+    }
+
+    public boolean initPollyServiceInterface() {
+        // config file correct
+        boolean configOK = true;
+
+        // Validate access key
+        if (PollyClientConfig.accessKey == null) {
```
[the remark this answers is not in the document]
Comment (hillmanr) [reply addons/voice/org.openhab.voice.pollytts/src/main/java/org/openhab/voice/pollytts/internal/cloudapi/PollyClientConfig.java]: Interesting point. I wrote this before I knew about the config.xml file. I'll assume the config is initialized and go with some code reduction.

File: addons/voice/org.openhab.voice.pollytts/ESH-INF/config/config.xml [hunk 15]
```diff
@@ -0,0 +1,47 @@
...
+	<config-description uri="voice:pollytts">
+		<parameter name="accessKey" type="text" required="true">
+			<label>Access Key</label>
+			<description>The Access Key to get access. You need to register to get a key.</description>
+		</parameter> 
+		<parameter name="secretKey" type="text" required="true">
+			<label>Secret Key</label>
+			<description>The Secret Key to get access. You need to register to get a key.</description>
+		</parameter>
+		<parameter name="serviceRegion" type="text" required="true">
```
[the remark this answers is not in the document]
Comment (hillmanr) [reply addons/voice/org.openhab.voice.pollytts/ESH-INF/config/config.xml]: ok

File: addons/voice/org.openhab.voice.pollytts/README.md [hunk 16]
```diff
@@ -0,0 +1,99 @@
+# Pollytts Text-to-Speech
+
+## Overview
+
+PollyTTS is a TTS voice service for Openhab 2.x utilizing a Internet based TTS service provided by amazon called Polly. There are five servers set in various geographic locations. It requires a set of API keys provided by Amazon to get access to the service. Provides multiple languages and multiple voices for each language. The free tier includes 5 million characters per month for the first 12 months. Then it's $4 for a million characters after the one year.
```
[the remark this answers is not in the document]
Comment (hillmanr) [reply addons/voice/org.openhab.voice.pollytts/README.md]: ok

**Comment by martinvw** [comment]
It would be great to squash the commits now, because all/most seem to lack a `Signed-off-by` (to certify that you wrote it or otherwise have the right to pass it on as an open-source), see http://docs.openhab.org/developers/contributing/contributing.html#sign-your-work

For squashing / rebasing see:
http://gitready.com/advanced/2009/02/10/squashing-commits-with-rebase.html

Please note that because of the way you merged in the commits it will be hard / impossible to do this easily. Maybe it's more easy to rename your branch locally and create a single new commit based on the current master where you commit the complete current state of the binding in one commit including the signed-off-by message.

**Comment by kubawolanin** [comment]
Thank you @hillmanr for fixing all the issues and code review remarks! Looking forward to see your add-on in the 2.2 distro! :)

**Comment by hillmanr** [comment]
Trying to move to clean signed commit

**Pull request closed without merging**